File: nQueenBacktracking.py
import time

def prenable(reines, ligne):
    if ligne in reines:
        return True
    nb = len(reines)
    diagPlus = [reines[i]+i for i in range(nb)]
    diagMoins = [reines[i]-i for i in range(nb)]
    if ligne+nb in diagPlus or ligne-nb in diagMoins:
        return True
    return False

tps1 = time.time()

n = 10
sols = [[]]
for c in range(1,n+1):
    newSols = []
    for solution in sols:
        for l in range(1,n+1):
            if not prenable(solution,l):
                newSol = solution.copy()
                newSol.append(l)
                newSols.append(newSol)
    sols = newSols
    #print(sols)
    
tps2 = time.time()
tmp2 = tps2 - tps1

print("time : ",tmp1)
print(len(sols)," solutions")
print(sols)

File: nQueenBackTracking.go
package main

import (
	"fmt"
)

func In(a int, list []int) bool{
    for _, b := range list {
        if b == a {
            return true
        }
    }
    return false
}

func prenable(reines []int, ligne int) bool{
	if In(ligne, reines){
        return true
	}
	nb := len(reines)
    diagPlus := []int{}
	diagMoins := []int{}
	for i := 0; i < nb; i++ {
		diagPlus = append(diagPlus, reines[i]+i)
		diagMoins = append(diagMoins, reines[i]-i)
	}
    return In(ligne+nb, diagPlus) || In(ligne-nb, diagMoins)
}

func swaptab(a *[][]int, b *[][]int) {
    c := *a
    *a = *b
    *b = c
}

func nilTab2() [][]int{
    vide := [][]int{}
    return vide
}

func main() {
	n := 12
    sols := [][]int{}
	list := []int{}
    for k := 0; k < n; k++{
		list = append(list, k)
        sols = append(sols, list)
		list = nil
    }
	newSol := []int{}
	newSols := [][]int{}
    for c := 1; c < n; c++{
		newSols = nilTab2()
		for solution := 0; solution < len(sols); solution++ {
			for l := 0; l < n; l++{
				if !prenable(sols[solution], l){
					newSol = make([]int, c)
					copy(newSol, sols[solution])
					newSol = append(newSol, l)
					newSols = append(newSols, newSol)
				}
			}
		}
		swaptab(&sols, &newSols)
    }
	fmt.Println(len(sols)," solutions")
	fmt.Println(sols)
}